CORE-6008: FakeCorda PersistenceService injected into flows

LizR3 opened the pull request:
Uses HSQLDB in-memory by default
Registered with the FiberMock to keep connections alive & allow access for testing
Made example CorDapp a bit more sane

LizR3 committed:
CORE-6008: CordaMock PersistenceService injected into flows

Uses HSQLDB in-memory by default
Registered with the FiberMock to keep connections alive & allow access for testing
Made example CorDapp a bit more sane

LizR3 committed:
CORE-6008: Removed unused import

LizR3 committed:
CORE-6008: Fixed example tests

driessamyn commented on gradle.properties:
```diff
@@ -96,6 +96,9 @@ osgiTestJunit5Version=1.1.0
 slingVersion=3.3.0
 
+# CorDapp test utils dependency versions
+jpaVersion=2.2
```
This is already defined in https://github.com/corda/corda-api/blob/e4473e2e17975405ff31ec5f99b6b55bf63f933e/gradle.properties#L45
and provided by `platform("net.corda:corda-api:$cordaApiVersion")` 

Example: https://github.com/corda/flow-worker/blob/feef661738a2fdf750f00671ec2400b5bdcd929b/libs/db/db-orm/build.gradle

LizR3 replied:
Thanks, will take it out.

LizR3 committed:
CORE-6008: Removed explicit JPA dependency

LizR3 committed:
CORE-6008: Also removed reference to JPA

JamesHR3 commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,124 @@
+
+    override fun <T : Any> merge(entities: List<T>): List<T> {
+        emf.transaction { em ->
+            return entities.map { em.merge(it) }
```
Not really a comment on this, but a note that as of today the list versions of these APIs isn't implemented in Corda. That should be done prior to DP2 so it shouldn't be an issue, but be aware that if you try this out now it won't succeed.

LizR3 replied:
Thanks! I spotted that `query` wasn't implemented but missed those. I'll leave them in for now but will make a story for myself to revisit and refresh what I'm doing closer to time.

JamesHR3 commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,124 @@
+import javax.persistence.EntityManager
+import javax.persistence.EntityManagerFactory
+
+class DbPersistenceService(x500 : MemberX500Name) : PersistenceService {
```
If any of the APIs in this fail, what will they throw? I think the actual flow service implementation will throw a `CordaPersistenceException`, which is exposed on the API so I think it is possible for flow code to catch it. (That's not very well tested at the moment!)

LizR3 replied:
I can do that now for sure; there's a TODO in the README.md to make all the exceptions match.

JamesHR3 reviewed (commented):
I've mostly looked at the `DbPersistenceService`.  Not too much to say, but I am a little concerned we might not properly represent what will happen in error scenarios.

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+## Instance vs Class upload
+
+The CordaMock has two methods of uploading responder flows:
+- via a class, which will be constructed when a response flow is initialized
```
```suggestion
- via a class, which will be constructed when a response flow is initialized.
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+- via a class, which will be constructed when a response flow is initialized
+- via an instance, which must be uploaded against a protocol.
+
+Uploading an instance allows flows to be constructed containing other mocks, injected logic, etc.. It also allows
```
```suggestion
Uploading an instance allows flows to be constructed containing other mocks, injected logic, etc. It also allows
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+
+The CordaMock has several components which can also be used independently:
+
+- A `CordaFlowChecker` which checks your flow for a default constructor and required Corda annotations
```
```suggestion
- A `CordaFlowChecker` which checks your flow for a default constructor and required Corda annotations.
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+The CordaMock has several components which can also be used independently:
+
+- A `CordaFlowChecker` which checks your flow for a default constructor and required Corda annotations
+- A `SimpleJSonMarshallingService` which can be used to convert objects to JSON and vice-versa
```
```suggestion
- A `SimpleJSonMarshallingService` which can be used to convert objects to JSON and vice-versa.
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+which allows the `RPCRequestData` to be easily constructed. There are three different construction
+methods available:
+
+- A Json-formatted string, as you would submit with `curl`:
```
```suggestion
- A JSON-formatted string, as you would submit with `curl`:
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+
+- A `CordaFlowChecker` which checks your flow for a default constructor and required Corda annotations
+- A `SimpleJSonMarshallingService` which can be used to convert objects to JSON and vice-versa
+- A `PassThroughFlowEngine` which will call any `SubFlows` you give to it
```
```suggestion
- A `PassThroughFlowEngine` which will call any `SubFlows` you give to it.
```

driessamyn commented on cordapp-test-utils/README.md:
```diff
@@ -0,0 +1,109 @@
+- A `CordaFlowChecker` which checks your flow for a default constructor and required Corda annotations
+- A `SimpleJSonMarshallingService` which can be used to convert objects to JSON and vice-versa
+- A `PassThroughFlowEngine` which will call any `SubFlows` you give to it
+- A `DbPersistenceService` using an in-memory HSQLDB
```
```suggestion
- A `DbPersistenceService` using an in-memory HSQLDB.
```

driessamyn commented on cordapp-test-utils/example-app/src/integrationTest/kotlin/net/cordacon/example/RollCallTest.kt:
```diff
@@ -50,7 +50,13 @@ class RollCallTest {
         """.trimIndent().replace("\n", System.lineSeparator())))
+
+        // And Ferris Bueller's absence should have been persisted
+        val persistence = corda.getPersistenceServiceFor(teacher)
```
Should there be a concept of a Node (teacher)? A Vnode in C5 is a combination of the X500 and a group ID. For now, it probably doesn't matter too much as I don't _think_ we support x-network operations, but this would technically be possible.
I also wonder if we have a `Node` or `VirtualNode` data structure it may be clearer what it is as X500 names could exist outside of a Corda Node.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/BaseFiberMock.kt:
```diff
@@ -1,15 +1,17 @@
  * Registers, and looks up, responder flows via their protocol.
  */
-class ProtocolLookUpFiberMock : FiberMock {
+class BaseFiberMock : FiberMock {
```
I think this is more of a Stub or Fake than a Mock. 
If it was a Mock, it'd be great to be able to test behaviour. The examples show asserting state at the end of the flow, and stubs can be very useful, but mocks are often more powerful as you can verify behaviour. E.g. `did it invoke a delete` rather than `is the end result it has deleted`. 
Supporting this with mocks would also mean you may not need a HSQL implementation, which has its own limitations. For example some queries will behave differently in Postgres vs. HSQL, and/or we may want to support native queries at some point for example. Somebody testing their CordApp is generally only interested in how their code is interacting with the API, not the result of the interaction.

driessamyn commented on cordapp-test-utils/example-app/src/main/kotlin/net/cordacon/example/RollCallFlow.kt:
```diff
@@ -12,10 +12,15 @@ import net.corda.v5.application.marshalling.JsonMarshallingService
 import net.corda.v5.base.annotations.Suspendable
 import net.corda.v5.base.types.MemberX500Name
 import net.corda.v5.base.util.contextLogger
+import java.util.*
```
I thought we disallowed wildcard imports? I'm surprised detekt hasn't moaned about it.

LizR3 replied:
Me too, also I thought I had disabled it in IntelliJ... will fix.

driessamyn commented on cordapp-test-utils/example-app/src/main/kotlin/net/cordacon/example/RollCallFlow.kt:
```diff
@@ -143,4 +158,17 @@ data class RollCallInitiationRequest(val recipientsX500: List<String>)
+
+@CordaSerializable
+@Entity
+data class AbsenceRecordEntity(
```
This should probably be in a contract CPK. This will work technically, and it doesn't matter probably till we have a ledger, but I thought our guidance was to have custom schemas in a separate CPK.

Possibly more important, this app doesn't seem to have a liquibase file. If we go through the effort of creating a HSQL DB, then we may as well do the DB migration as well as that is possibly one of the only good reasons to have a real DB, at least it validates the migration files, which are very hard to otherwise test without doing an actual deployment.

LizR3 committed:
CORE-6008: DbPersistenceService throws CordaPersistenceException

driessamyn commented on cordapp-test-utils/test-utils/build.gradle:
```diff
@@ -11,6 +11,9 @@ dependencies {
     implementation "org.slf4j:slf4j-api:$slf4jVersion"
 
+    implementation "org.hibernate:hibernate-core:$hibernateVersion"
+    implementation "org.hsqldb:hsqldb:$hsqldbVersion"
```
I think you only need this as a runtime dependency.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/CordaMock.kt:
```diff
@@ -34,8 +35,8 @@ import net.corda.v5.base.types.MemberX500Name
-    private val injector: FlowServicesInjector = SensibleServicesInjector(),
-    private val fiberMock: FiberMock = ProtocolLookUpFiberMock()
+    private val fiberMock: FiberMock = BaseFiberMock(),
+    private val injector: FlowServicesInjector = SensibleServicesInjector()
```
Not a big fan of calling something `SensibleServicesInjector` as it sort of implies the real one isn't a sensible implementation.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/BaseFiberMock.kt:
```diff
@@ -43,10 +45,19 @@ class ProtocolLookUpFiberMock : FiberMock {
         }
     }
 
+    override fun registerPersistenceService(x500: MemberX500Name, persistence: PersistenceService) {
```
I think you use `member` elsewhere and here `x500`. Probably worth keeping consistent, although I think node/vnode would be more suitable (along with a vnode type)

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/BaseFiberMock.kt:
```diff
@@ -43,10 +45,19 @@ class ProtocolLookUpFiberMock : FiberMock {
         return nodeInstances[member]?.get(protocol)
     }
 
+    override fun getPersistenceService(x500: MemberX500Name): PersistenceService {
```
As above x500/member

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/FiberMock.kt:
```diff
@@ -1,14 +1,14 @@
-}
-
-interface ProtocolLookUp {
+    fun registerPersistenceService(x500: MemberX500Name, persistence: PersistenceService)
```
As above x500/member

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/FiberMock.kt:
```diff
@@ -1,14 +1,14 @@
+    fun registerPersistenceService(x500: MemberX500Name, persistence: PersistenceService)
     fun lookUpResponderClass(member: MemberX500Name, protocol: String): Class<out ResponderFlow>?
     fun lookUpResponderInstance(member: MemberX500Name, protocol: String): ResponderFlow?
+    fun getPersistenceService(x500: MemberX500Name): PersistenceService
```
As above x500/member

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/internal/JpaPersistenceUnitInfo.kt:
```diff
@@ -0,0 +1,62 @@
+import javax.sql.DataSource
+
+@Suppress("TooManyFunctions")
+class JpaPersistenceUnitInfo : PersistenceUnitInfo {
```
FYI, we already have one of these: https://github.com/corda/flow-worker/blob/6d23693274f7eaa61f4f9421a22d65bf46e79433/libs/db/db-orm-impl/src/main/kotlin/net/corda/orm/impl/CustomPersistenceUnitInfo.kt

Some of the other classes in `db-core` and `db-orm` (and `db-admin` if you add DB migration) may be applicable, but I'm not sure it's worth taking a reference or re-implementing. Definitely worth thinking about and coming up with pros/cons.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,139 @@
+
+class DbPersistenceService(x500 : MemberX500Name) : PersistenceService {
+
+    private val emf = createEntityManagerFactory(x500)
```
EMFs need to be closed when you're done with them.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+inline fun <R> EntityManagerFactory.transaction(block: (EntityManager) -> R): R {
```
A lot of duplication with https://github.com/corda/flow-worker/blob/6d23693274f7eaa61f4f9421a22d65bf46e79433/libs/db/db-orm/src/main/kotlin/net/corda/orm/utils/EntityManagerUtils.kt

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,139 @@
+
+inline fun <R> EntityManagerFactory.transaction(block: (EntityManager) -> R): R {
+
+    val em = this.createEntityManager()
+    val t = em.transaction
+    t.begin()
+    return try {
+        block(em)
+    } catch (e: Exception) {
+        t.setRollbackOnly()
+        throw CordaPersistenceException(e.message ?: "Error in persistence", e)
+    } finally {
+        if (!t.rollbackOnly) {
+            t.commit()
+        } else {
+            t.rollback()
+        }
+    }
```
You need to do something like: 
```suggestion
this.createEntityManager().use { em ->
    val t = em.transaction
    t.begin()
    return try {
        block(em)
    } catch (e: Exception) {
        t.setRollbackOnly()
        throw CordaPersistenceException(e.message ?: "Error in persistence", e)
    } finally {
        if (!t.rollbackOnly) {
            t.commit()
        } else {
            t.rollback()
        }
    }
}
```

To ensure the EM is closed when you're finished.
As it stands, this will leak.

driessamyn reviewed (changes requested):
Most of these comments are very nitpicky, feel free to ignore, but there are some things like EMs and EMFs need to be closed when you're finished, which I feel should be tidied up before this is merged.

I think calling this a mock isn't entirely correct as it's a stub, and I wonder if a mock would be more useful in certain cases, but realise this is debatable.
However, if we have this as a stub, and go through the effort of using a real DB (which has it's own downsides), then I think at least we should integration DB migrations as well as that is one of the few things I can think of that would add value when using a real DB rather than real mocks. DB migrations are otherwise very hard to test before doing an actual deployment.

I haven't reviewed the CordApp.

LizR3 committed:
CORE-6008: Apply doc fixes from code review

Co-authored-by: Dries Samyn <dries.samyn@r3.com>

LizR3 committed:
CORE-6008: Renaming from code review comments
CORE-6008: Closing entity managers & factories

LizR3 commented:
All the review comments are now addressed, or have stories raised against them, except for the pros / cons discussion about using corda-runtime-os's internal classes. I'm not keen to do that at the moment as I think we have slightly different requirements, and our classes do seem to keep diverging, but I'll put together a list of pros / cons and post it to the channel.

driessamyn commented on cordapp-test-utils/test-utils/src/main/kotlin/net/corda/testutils/services/DbPersistenceService.kt:
```diff
@@ -0,0 +1,144 @@
+        } else {
+            t.rollback()
+        }
+        em.close()
```
This will not be closed when commit or rollback throws.

Why not use `em.use {...}` ?

LizR3 committed:
CORE-6008: Closeable DB, CordaMock -> FakeCorda

Also renamed FiberFake to FakeFiber to match
Added missing tests for DefaultServicesInjector
Ensured only one PersistenceService per member

LizR3 committed:
CORE-6008: Adding EntityManager.use

Currently copying Corda's code so as not to leak internals unnecessarily
Fixed in JPA 3.1.0 (we are on 2.2)

LizR3 committed:
CORE-6008: Small fixes

driessamyn commented on cordapp-test-utils/test-utils/src/test/kotlin/net/corda/testutils/FakeCordaTest.kt:
```diff
@@ -21,34 +21,34 @@ import org.mockito.kotlin.times
         // Given a mock flow checker in our mock Corda network
         val flowChecker = mock<FlowChecker>()
-        val corda = CordaMock(flowChecker)
+        val corda = FakeCorda(flowChecker)
```
Bit of a detail, but now that FakeCorda is closeable (good), you should close these even in the test.
I think the Kotlion idiomatic way would be to do:

```kotlin
FakeCorda(flowChecker).use { corda ->
 ...
}
```

driessamyn commented on cordapp-test-utils/test-utils/src/test/kotlin/net/corda/testutils/internal/DefaultServicesInjectorTest.kt:
```diff
@@ -0,0 +1,28 @@
+
+        // With some helpful classes to use in services
+        val member = MemberX500Name.parse("CN=IRunCorDapps, OU=Application, O=R3, L=London, C=GB")
+        val fiber = BaseFakeFiber()
```
Similar to the above, but possibly more important as this is an "example", the fiber needs to be closed now that it is Closeable.

So:

```kotlin
BasFakeFiber().use { fiber ->
...
}
```

driessamyn commented on cordapp-test-utils/test-utils/src/test/kotlin/net/corda/testutils/services/DbPersistenceServiceTest.kt:
```diff
@@ -0,0 +1,161 @@
+    @Test
+    fun `should connect using hsqldb as default`() {
+        // Given a persistence service for a member
+        val persistence = DbPersistenceService(member)
```
Same comment as above about closing things. This case is probably the most important one as it would clean up the DB when the test is done.

driessamyn reviewed (approved):
I just left a couple more comments about closing things. Up to you if you want to make the change before merging or not. If not, I suggest doing this soon after as it sets a good example to close up resources when you're done with them.

ronanbrowne reviewed (approved):
LGTM - for build changes - merging on @LizR3  request
